### PR TITLE
Multiple Merchants for Syndicates

### PR DESCRIPTION
**Pull request by mabbus**
- synd cant log in anymore
- login for synd/merchants
- add business role on attach merchant to business

### CODE REVIEW

File: app/Deal.php [hunk 1]
```diff
@@ -418,6 +418,10 @@ public function getLastMerchantPayAttribute()
 
     public function getEstimatedPayoffDateAttribute ()
     {
+        if ($this->startDate === 0){
```
Comment (lexod) [inline app/Deal.php:421]: there is already a fix for it
https://github.com/mabbus/enablement-api/pull/88/files#diff-5f2b341af34f0cda1e17893bbbd53dbb9986a11e758fb5869c53c65a26c26316R426

File: app/Http/Tasks/Business/AddMerchantToBusinessTask.php [hunk 2]
```diff
@@ -13,19 +17,31 @@ class AddMerchantToBusinessTask implements TaskInterface
 
     /**
      * AddMerchantToBusinessTask constructor.
-     * @param Business $business
+     * @param Model $entity
      * @param array $data
      */
-    public function __construct(Business $business, array $data)
+    public function __construct(Model $entity, array $data, $role = false )
```
Comment (lexod) [inline app/Http/Tasks/Business/AddMerchantToBusinessTask.php:23]: should be `public function __construct(Model $entity, array $data, Role $role = null )`

File: app/Http/Tasks/Business/AddMerchantToBusinessTask.php [hunk 3]
```diff
@@ -13,19 +17,31 @@ class AddMerchantToBusinessTask implements TaskInterface
...
     {
-        $this->business = $business;
+        $this->entity = $entity;
```
Comment (lexod) [inline app/Http/Tasks/Business/AddMerchantToBusinessTask.php:25]: needs a check for types of entity here

File: app/Http/Tasks/Business/AddMerchantToBusinessTask.php [hunk 4]
```diff
@@ -13,19 +17,31 @@ class AddMerchantToBusinessTask implements TaskInterface
...
+        try{
+            $this->entity->merchants()->attach($this->data['merchant_id']);
+            $this->entity->save();
+            $merchant = Merchant::find($this->data['merchant_id']);
+            if($this->role){
+                $merchant->user->roles()->attach($this->role);
+            }
+            DB::commit();
+            return $this->entity->refresh()->merchants;
+        } catch (\Exception $e) {
```
Comment (lexod) [inline app/Http/Tasks/Business/AddMerchantToBusinessTask.php:42]: should be `\Throwable`

File: app/Http/Tasks/Business/AddMerchantToBusinessTask.php [hunk 5]
```diff
@@ -13,19 +17,31 @@ class AddMerchantToBusinessTask implements TaskInterface
...
+            return response()->json($e);
```
Comment (lexod) [inline app/Http/Tasks/Business/AddMerchantToBusinessTask.php:43]: task cannot return JSON response it's a responsibility of the Controller.
Throw exception instead.

File: app/Http/Controllers/Web/SyndicateDealsController.php [hunk 6]
```diff
@@ -49,7 +49,7 @@ public function show(GetDeal $request, Deal $deal) {
                 'deal' => $syndDeal,
                 'transactions' => $syndDeal->transactions,
                 'payments' => $payments,
-                'syndicate' => $user->syndicate
+                'syndicate' => $user->merchant->syndicate->first()
```
Comment (lexod) [inline app/Http/Controllers/Web/SyndicateDealsController.php:52]: use $user->load()

File: app/Http/Controllers/Web/SyndicateDealsController.php [hunk 7]
```diff
@@ -29,13 +29,13 @@ public function index()
...
         $data = $request->validated();
-        $syndicate = Auth::user()->syndicate;
+        $syndicate = Auth::user()->merchant->syndicate->first();
         return response()->json((new SyndicateGetDeals($data, $syndicate))->run());
     }
 
     public function show(GetDeal $request, Deal $deal) {
         $user = Auth::user();
-        $syndDeal = (new SyndicateGetDeal($deal, $user->syndicate))->run();
+        $syndDeal = (new SyndicateGetDeal($deal, $user->merchant->syndicate->first()))->run();
```
Comment (lexod) [inline app/Http/Controllers/Web/SyndicateDealsController.php:38]: this line duplicates line 52

File: app/Http/Tasks/Business/AddMerchantToBusinessTask.php [hunk 8]
```diff
@@ -13,19 +17,31 @@ class AddMerchantToBusinessTask implements TaskInterface
...
     public function run()
     {
-        $this->business->merchants()->attach($this->data['merchant_id']);
-        $this->business->save();
-        return $this->business->refresh()->merchants;
+        DB::beginTransaction();
+        try{
+            $this->entity->merchants()->attach($this->data['merchant_id']);
+            $this->entity->save();
+            $merchant = Merchant::find($this->data['merchant_id']);
```
Comment (lexod) [inline app/Http/Tasks/Business/AddMerchantToBusinessTask.php:36]: use firstOrFail() if merchant doesn't exist

File: app/Http/Tasks/Syndicate/CreateSyndicate.php [hunk 9]
```diff
@@ -19,34 +18,25 @@ public function __construct(array $data)
...
-                return $syndicate;
-            } catch (\Throwable $e) {
-                DB::rollback();
-                throw $e;
-            }
-        };
+        DB::beginTransaction();
+        try {
+            /*
+            $user = new User($this->data);
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/CreateSyndicate.php:24]: get rid of the comments.
we have git for it

File: app/Http/Tasks/Syndicate/CreateSyndicate.php [hunk 10]
```diff
@@ -19,34 +18,25 @@ public function __construct(array $data)
...
-                $syndicate = new Syndicate($this->data);
-                $syndicate->save();
-                DB::commit();
-                return $syndicate;
-            } catch (\Throwable $e) {
-                DB::rollback();
-                throw $e;
-            }
-        };
+        DB::beginTransaction();
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/CreateSyndicate.php:21]: it's a single table.
no need for transactions

File: app/Http/Tasks/Syndicate/CreateSyndicateTask.php [hunk 11]
```diff
@@ -21,8 +21,8 @@ public function __construct( User $account, array $profileData)
     public function run()
     {
         $syndicate = new Syndicate($this->profileData);
-        $syndicate->user()->associate($this->account);
-        $syndicate->user()->roles()->attach(Role::syndicate());
+        //$syndicate->user()->associate($this->account);
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/CreateSyndicateTask.php:24]: get rid of the comments.
we have git for it

File: app/Http/Tasks/Syndicate/CreateSyndicate.php [hunk 12]
```diff
@@ -6,7 +6,6 @@
 use App\Syndicate;
 use App\User;
 use Illuminate\Support\Facades\DB;
-use Illuminate\Support\Facades\Hash;
 
 class CreateSyndicate implements TaskInterface
```
Comment (lexod) [inline app/Http/Tasks/Syndicate/CreateSyndicate.php:10]: what is the difference between this task and `app/Http/Tasks/Syndicate/CreateSyndicateTask.php` ?

File: app/Merchant.php [hunk 13]
```diff
@@ -58,6 +58,11 @@ public function businesses()
         return $this->belongsToMany(Business::class, 'merchants_has_businesses');
     }
 
+    public function syndicate()
```
Comment (lexod) [inline app/Merchant.php:61]: if it belongs to many - should be `syndicates`

File: database/migrations/2020_08_05_195543_create_roles_table.php [hunk 14]
```diff
@@ -14,9 +14,7 @@ private function getData()
             \App\Role::ROOT,
             \App\Role::ADMIN,
             \App\Role::SYNDICATE,
-            \App\Role::MERCHANT,
-            \App\Role::BROKER,
-            \App\Role::HOST,
+            \App\Role::BUSINESS
```
Comment (lexod) [inline database/migrations/2020_08_05_195543_create_roles_table.php:17]: very bad idea to edit old migrations
create a new one instead.
don't forget to update existing roles and users mappings

File: app/Role.php [hunk 15]
```diff
@@ -13,15 +13,15 @@ class Role extends Model
     const HOST = 'host';
     const SYNDICATE = 'syndicate';
     const BROKER = 'broker';
-    const MERCHANT = 'merchant';
+    const BUSINESS = 'business';
```
Comment (lexod) [inline app/Role.php:16]: sync up users, roles and mapping table

File: resources/views/employees/syndicates/show.blade.php [hunk 16]
```diff
@@ -15,6 +16,9 @@
                 errors: [],
                 hasErrors: [],
                 routing_number: '',
+                merchant_id: '',
```
Comment (lexod) [inline resources/views/employees/syndicates/show.blade.php:19]: will have to change this whole thing to a component

**Review by lexod — CHANGES_REQUESTED** [review]

File: app/Deal.php (hunk 1 above)
Comment (mabbus) [reply app/Deal.php:421]: this should have been it's own pr then

File: app/Http/Tasks/Syndicate/CreateSyndicate.php (hunk 12 above)
Comment (mabbus) [reply app/Http/Tasks/Syndicate/CreateSyndicate.php:10]: I think one i used by API and one is used by web

File: app/Merchant.php (hunk 13 above)
Comment (mabbus) [reply app/Merchant.php:61]: belongs to 1

File: app/Http/Controllers/Web/EmployeesSyndicatesController.php [hunk 17]
```diff
@@ -70,7 +75,12 @@ public function show(GetSyndicate $request, Syndicate $syndicate) {
             $dealsTotal->payment = ($dealsTotal->amount / $dealsTotal->terms) ?? 1;
         }
 
-        return view('employees.syndicates.show', ['title' => 'Syndicates | ' . $syndicate->business_name, 'syndicate' => $syndicate, 'dealsTotal' => $dealsTotal, 'deals' => $deals ]);
+        return view('employees.syndicates.show', [
+            'title' => 'Syndicates | ' . $syndicate->business_name,
```
Comment (lexod) [inline app/Http/Controllers/Web/EmployeesSyndicatesController.php:79]: `"Syndicates |  $syndicate->business_name"`

**Review by lexod — APPROVED** [review]

**Pull request merged**